### DISCUSSION TITLE
Constant high I/O wait caused by Photoprism

### DISCUSSION THREAD

**Discussion by donateur** [photoprism/photoprism#3603] [Q&A] [answered]
Hello,

I am struggling to diagnose a problem and would love any suggestions please!

I noticed a ticking sound from my Linux media PC recently and thought that the hard disks were failing.  However when I checked the OS I found that one CPU was constantly stuck in I/O wait.  Even though I cannot identify the process causing this (see the attached screenshot, where htop is sorted by combined Read/Write and the network is quiet) I know it is related to Photoprism because when I stop photoprism (all of its containers) the I/O wait instantly disappears.  If I re-start photoprism, the I/O wait is back, exactly as visible in the picture.

![photoprism-iowait](https://github.com/photoprism/photoprism/assets/12554885/e771ef16-4ce9-4c7a-a1ef-af5d55519ea7)

Aside from the I/O wait problem, the PC is almost idle, with a load average of 0.07, plenty of free resources.  While it does run other software, nothing else is doing much I/O at the times I've been trying to diagnose this problem.

* I'm running Fedora Linux version 37, with moby (forked open source Docker).
* Photoprism containers, cache, database and sidecare files are stored on an SSD while the photos are stored on a mirrored BTRFS volume across two HDDs.  The first volume (on the SSD) is 75% full while the latter has hundreds of gigabytes free.
* There are no SELinux denials or any obvious log messages at the OS level or within the photoprism container (using docker logs etc)
* other diagnostic tools show the same I/O wait, but don't show what's causing it
* There is other software on this PC but I've always tried to diagnose this when it is idle
* There are no errors in the Photoprism UI and Photoprism appears to work as normal

Where should I start to diagnose this please?

**Comment by donateur** [comment]
I've spent more time looking and while other tools show I/O wait on the CPU, they also don't identify the process causing it!  I spent more time digging through the container logs associated with photoprism but can't see any errors or problems either.

**Comment by donateur** [comment]
One of the maintainers of 'htop' kindly offered help, so I now know a little more.  The problem is related to some Photoprism processes being affected by a "Pressure Stall" relating to IO [1].  He suggested that this can happen when containers have too few resources and are trying to 'swap out' memory.  It's not an actual lack of resources as you can see from the screenshot.

While I had changed the example docker-compose.yml to reduce PHOTOPRISM_WORKERS to 2 from the default of 4, I changed this back, down'ed and up'ed with docker compose but the problem persists.

I have not changed any other resource related settings, all is default.  Any ideas please?

[1]: https://www.kernel.org/doc/html/latest/accounting/psi.html

**Reply by lastzero** [reply]
I'm unfortunately busy with a lot of other tasks right now, so can't offer my help to dive into this. Maybe someone from our community can?

**Comment by donateur** [comment]
> I'm unfortunately busy with a lot of other tasks right now, so can't offer my help to dive into this. Maybe someone from our community can?

Thank you, this isn't urgent.  I was originally concerned that it might be causing damage to my hard disks, but it's only effecting Photoprism.  On the positive side when this is resolved Photoprism should perform better on my system!

**Reply by lastzero** [reply]
Happy to fix it if possible and I know what exactly must be changed. Could it be caused by a file that wasn't closed after reading / writing or one of the libraries / dependencies we use? For example Sqlite or TensorFlow?

**Reply by lastzero** [reply]
You may also want to check the `entrypoint.sh` script we use, for example:

https://github.com/photoprism/photoprism/blob/b194188e91824b62e5fa4307151144b09066c391/scripts/dist/entrypoint.sh#L123-L124

**Reply by donateur** [reply]
Thanks again!

I don't have TensorFlow enabled.  In the other issue someone has helpfully suggested it may be a cosmetic issue caused by the kernel.  Here's a kernel commit to fix a similar sounding symptom from the 24th July:
> A previous commit made all cqring waits marked as iowait, as a way to
improve performance for short schedules with pending IO. However, for
use cases that have a special reaper thread that does nothing but
wait on events on the ring, this causes a cosmetic issue where we
know have one core marked as being "busy" with 100% iowait.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7b72d661f1f2f950ab8c12de7e2bc48bdac8ed69

Does that mean much to you?  Do you know if any of the software in photoprism uses the io_uring system kernel call?  Apparently it is used for for storage device asynchronous I/O operations.
This is above my head!

I'm running the older release of Fedora but should get new kernels pretty quickly.  The same helpful person has suggested that the fix is backported in to kernel 6.4.8 which I should have in a week or two.

**Reply by lastzero** [reply]
Wouldn't be surprised as modern kernels include all kinds of tweaks. You often can't trust the numbers. I suggest you update the base distribution before looking any further!

**Reply by donateur** [reply, marked as answer]
I just checked and a new kernel (6.4.9) was available.  I've rebooted now and the problem is resolved.  It seems most likely it was a cosmetic issue related to the io_uring syscall and was resolved by that kernel patch above.  It's probably worth being aware of this in case other people come across it.

Thank you again for your help and this great software @lastzero
